pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -278,49 +515,233 @@ func getRequirements(solveNode map[string]interface{}) ([]Requirement, error) {
 //	    }
 //	  }
 //	}
-func getSolveNode(expression map[string]interface{}) (map[string]interface{}, error) {
-	solveNode, err := getFuncNode(expression, SolveFuncName)
-	if err == nil {
-		return solveNode, nil
+func (e *BuildExpression) getSolveNode() *Ap {
+	for _, a := range e.Let.Assignments {
+		if a.Value.Ap == nil {
+			continue
+		}
+
+		if a.Value.Ap.Name == SolveFuncName || a.Value.Ap.Name == SolveLegacyFuncName {
+			return a.Value.Ap
+		}
 	}
-	if !errors.Is(err, funcNodeNotFoundError) {
-		return nil, errs.Wrap(err, "Could not get solve node")
+
+	return nil
+}
+
+// Update updates the BuildExpression's requirements based on the operation and requirement.
+func (e *BuildExpression) Update(operation Operation, requirement Requirement, timestamp strfmt.DateTime) error {
+	var err error
+	switch operation {
+	case OperationAdded:
+		err = e.addRequirement(requirement)
+	case OperationRemoved:
+		err = e.removeRequirement(requirement)
+	case OperationUpdated:
+		err = e.updateRequirement(requirement)
+	default:
+		return errs.New("Unsupported operation")
+	}
+	if err != nil {
+		return errs.Wrap(err, "Could not update BuildExpression's requirements")
+	}
+
+	err = e.updateTimestamp(timestamp)
+	if err != nil {
+		return errs.Wrap(err, "Could not update BuildExpression's timestamp")
 	}
-	logging.Debug("Could not get solve node, trying solve legacy node")
 
-	return getFuncNode(expression, SolveLegacyFuncName)
+	return nil
 }
 
-// getFuncNode returns the node of the given function name from the build expression.
-// It returns an error if the function node is not found.
-// Currently, this function just recurses the build expression until it finds the function node
-// of the correct map[string]interface{} type.
-// It expects the JSON representation of the build expression to be formatted as follows:
-//
-//	{
-//	  "let": {
-//	    "runtime": {
-//	      "func_name": {
-//	      }
-//	    }
-//	  }
-//	}
-func getFuncNode(expression map[string]interface{}, funcName string) (map[string]interface{}, error) {
-	for k, v := range expression {
-		node, ok := v.(map[string]interface{})
-		if !ok {
+func (e *BuildExpression) addRequirement(requirement Requirement) error {
+	obj := []*Var{
+		{Name: RequirementNameKey, Value: &Value{Str: p.StrP(requirement.Name)}},
```

Comment:
Does Go auto-convert this struct to a pointer-to-struct? I've been doing:

```
[]*Var{
  &Var{Name: name, Value: value},
}
```

all this time. If what you're doing is fine, then I can save myself some typing!

Reply:
Yep, it's part of how composite literals work in Go. Some more information [here](https://go.dev/ref/spec#Composite_literals) if you're interested. There is an example near the end of that section that pertains to this usage.